### PR TITLE
Move tests

### LINKED ISSUE #982: About tests organization
Below is the first part of the path:

```
tester/src/test/scala/spinal/
```

The standard for unit tests would be:

```
PROJECT/src/test/scala/spinal/
```

So let's admit that (most) tests running simulations are integration tests, hence putting them in a separate `tester/` module. It also simplifies dependencies management for the day we want to use features released in `tester/` to test something from `core/` or `lib/`.

IMHO nothing to change here :+1:

---

```
tester/src/test/scala/spinal/tester/
                             ^^^^^^
```

Then there is the `tester/` folder. As it is about integration tests, it forces to import tested things (else using the same package as the test being run would bring in scope the stuff if the same package), making tests look more like examples, to better test the user's API, which seems to be a good thing.

IMHO nothing to change here :+1:

---

Then there are lots of tests with poor organization. Below is a suggestion:

* `drafts/` for worksheets (old things that could be removed IMHO), so that most contributors can forget them
* `path/to/item.scala` where the first element of the path is basically `core`, `lib`, etc. without being repeated.
* `issues/IssueXXX.scala` for bug fixes

For instance:

```
tester/src/test/scala/spinal/tester/scalatest/AhbLite3InterconnectTester.scala
```

tests:

```
lib/src/main/scala/spinal/lib/bus/amba3/ahblite/AhbLite3Interconnect.scala
```

So it would be moved to:

```
tester/src/test/scala/spinal/tester/lib/bus/amba3/ahblite/AhbLite3InterconnectTester.scala
```

Tests about different elements (elements being tested, not elements used to test other elements) would be in the deepest common element in the paths.

I think it is okay to have `App`s to test Scala syntax and types among scalatest-run tests.

This way it would be easier to find and add more tests.

### PR DESCRIPTION
**Pull request by numero-744**
TL;DR: gl hf

This PR moves tests to their standard location according to SBT. To do so:

* "test" dependencies between the different subprojects were added (`see build.sbt`)
* The things that I had to put common (or I thought I had to) were moved to `tester`, and its organization in packages should be discussed (eg: sim vs formal)
* I could not associate some tests with a specific item / file so for them, they are in `integration` folder of the appropriate sub-project.

Closes #982

### CODE REVIEW

**Comment by Readon** [comment]
So the dependency graph becomes complex though.

**Comment by numero-744** [comment]
I would say we just add one single rule: in tests, everything can access to everything :sweat_smile: 

Didn't change release (main) dependency graph.

File: core/src/main/scala/spinal/core/Vec.scala [hunk 1]
```diff
@@ -311,7 +311,7 @@ class VecBitwisePimper[T <: Data with BitwiseOp[T]](pimped : Vec[T]) extends Bit
 
   private def map2with(f: (T, T) => T)(other: Vec[T]): Vec[T] = {
     if (pimped.length != other.length)
-      SpinalError(s"Cannot apply a bitwize opration on vectors with different size (${pimped.length} vs ${other.length})")
+      SpinalError(s"Cannot apply a bitwize operation on vectors with different sizes (${pimped.length} vs ${other.length})")
```
Comment (Readon) [inline core/src/main/scala/spinal/core/Vec.scala:314]: Would you please commit this individually?

File: core/src/test/scala/integration/OperatorTester.scala [hunk 2]
```diff
@@ -1,26 +1,8 @@
...
- * version 3.0 of the License, or (at your option) any later version.
- *
- * This library is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
- * Lesser General Public License for more details.
- *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this library.
- */
```
Comment (Readon) [inline core/src/test/scala/integration/OperatorTester.scala:17]: Does this license should be leave unchanged? @Dolu1990

File: tester/src/test/scala/fixes/Issue963.scala [hunk 3]
```diff
@@ -1,10 +1,10 @@
-package spinal.tester.scalatest
+package fixes
```
Comment (Readon) [inline tester/src/test/scala/fixes/Issue963.scala:1]: This package name is a little bit weird.

File: tester/src/main/scala/spinal/tester/RepeatabilitySuite.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.io.File
+import org.apache.commons.io.FileUtils
+
+class RepeatabilitySuite extends AnyFunSuite {
+  var checkOutputHashCounter = 0
+  def checkOutputHash(gen: => Component): Unit = {
+    checkOutputHashCounter = checkOutputHashCounter + 1
+    var ref = ""
+    for (i <- 0 until 8) {
```
Comment (Readon) [inline tester/src/main/scala/spinal/tester/RepeatabilitySuite.scala:15]: What's the reason that constant 8 and 50MHz should be fixed?

File: tester/src/test/scala/fixes/Issue963.scala (hunk 3 above)
Comment (numero-744) [reply tester/src/test/scala/fixes/Issue963.scala:1]: I initially put `issues` but then found a PR test which had no issue attached to it, so this is why I have chosen this name: to fix bugs.

Maybe something like `bugs`, `ghbugs`, `bughistory`?

**Comment by kleinai** [comment]
I have a few concerns.  You should keep the refactoring to as little changes as necessary.  I understand package names and some imports will change.  I also feel we should squash these commits down to avoid creating a large commit log of "refractor moved X" that's annoying to look at. Listing all the files out in fewer commits is fine.

Also if we're moving files from core, lib, and test we need to make sure the licenses are compatible.  We probably haven't thought about this in the past but there's a considerable amount of code moving around now.

**Comment by numero-744** [comment]
I planned to have a semi-linear history for this, to ease reading the graph, but yes we can consider reducing the number of commits.

My concern is that sometimes there were several files testing the same thing; so I have put them together into a single file to match the SBT convention. In this case, mixing all the commits together might be annoying because it makes it harder to find from which file comes this added code chunk. If the blame shows that it comes from a commit with only:

- the file modified with adds
- and one deleted file

then, the code comes from this deleted file. See https://github.com/SpinalHDL/SpinalHDL/pull/1001#discussion_r1061736681

To sum up:

- history readability (that you mentioned)
- code traceability (that I mention)

What do you think about squashing commits which "only rename" files (even if there are some modifications)? So that we reduce the number of commits (maybe a lot?) while keeping good code traceability for when files are merged together.

Note: for now I'm working fast-forward only, I'll rebase and squash things when we agree on the code we get.

---

I think I have only moved files from `tester/scalatest` and I don't know what are the rules for this project, should check this. **As we'll publish `tester` (the structure has to be defined, for now I just have put everything in the same folder)** we should clearly define that it is the same license as `lib` IMO.

File: core/src/test/scala/integration/OperatorTester.scala (hunk 2 above)
Comment (Dolu1990) [reply core/src/test/scala/integration/OperatorTester.scala:17]: You can burn that license ^^

**Comment by Readon** [comment]
> I have a few concerns. You should keep the refactoring to as little changes as necessary. I understand package names and some imports will change. I also feel we should squash these commits down to avoid creating a large commit log of "refractor moved X" that's annoying to look at. Listing all the files out in fewer commits is fine.
> 
> Also if we're moving files from core, lib, and test we need to make sure the licenses are compatible. We probably haven't thought about this in the past but there's a considerable amount of code moving around now.

I am also think that one PR simply move the tester into right directory is enough.
More patchs can be commit by other PR, it's really too big to review now.

**Comment by Readon** [comment]
When would this be finished?^^
The existing test is really take too much time to finish.

**Comment by numero-744** [comment]
@Readon thanks for the ping, it is nice to know that what I've been working on will be useful :smiley: 

---

I think I won't be able to spend significant time on this PR until early Feb **but**:

I will need feedback for the following points (among the points of the TODO list in the PR description) so this PR can progress through discussion before I'm back on it:

- Find a better package name than `fixes` (see https://github.com/SpinalHDL/SpinalHDL/pull/1001#discussion_r1061738095)
- Check licenses (maybe delay that until later review?) (see https://github.com/SpinalHDL/SpinalHDL/pull/1001#issuecomment-1371472855)
- Define an organization of packages for [`tester/main`](https://github.com/numero-744/SpinalHDL/tree/move-tests/tester/src/main/scala/spinal)

When I'll be back on this PR I'll be able to:

- Squash "only rename" commits to reduce commits number as suggested in https://github.com/SpinalHDL/SpinalHDL/pull/1001#issuecomment-1371889929
- Apply suggestions according to the feedback (especially change `fixes` package name if needed and organize `tester/main`)
- Force-push that for review
- Remove draft status

**Comment by Readon** [comment]
> @Readon thanks for the ping, it is nice to know that what I've been working on will be useful 😃
> 
> I think I won't be able to spend significant time on this PR until early Feb **but**:
> 
> I will need feedback for the following points (among the points of the TODO list in the PR description) so this PR can progress through discussion before I'm back on it:
> 
> * Find a better package name than `fixes` (see [Move tests #1001 (comment)](https://github.com/SpinalHDL/SpinalHDL/pull/1001#discussion_r1061738095))
> * Check licenses (maybe delay that until later review?) (see [Move tests #1001 (comment)](https://github.com/SpinalHDL/SpinalHDL/pull/1001#issuecomment-1371472855))
> * Define an organization of packages for [`tester/main`](https://github.com/numero-744/SpinalHDL/tree/move-tests/tester/src/main/scala/spinal)
> 
> When I'll be back on this PR I'll be able to:
> 
> * Squash "only rename" commits to reduce commits number as suggested in [Move tests #1001 (comment)](https://github.com/SpinalHDL/SpinalHDL/pull/1001#issuecomment-1371889929)
> * Apply suggestions according to the feedback (especially change `fixes` package name if needed and organize `tester/main`)
> * Rebase to fix merge conflicts + move more stuff if needed + re-squash if needed
> * Force-push that for review
> * Remove draft status

Maybe the first step is just move those test into their corresponding directory to speed up the CI.
^^

**Comment by Dolu1990** [comment]
> thanks for the ping, it is nice to know that what I've been working on will be useful

Sure :D

The main thing to be carefull about, is to not let such massive PR too long in flight, as the more the time pass, the more conflict may popup.

> fixes 

could be regression ? or something else, as you feel it ^^

**Comment by andreasWallner** [comment]
Maybe another ping ;-)
To throw in another possible name: `spinal.issues` (since we are currently referring to issues in those?) 
Long term those might make more sense as `TestIssueXxx` tests with theirs respective modules I think, that would be the most likely place to look for them?

**Comment by Readon** [comment]
I think this have been done, so can we close this? @numero-744

**Comment by Dolu1990** [comment]
It isn't fully completted, but better to do it from upstream i think

**Comment by Dolu1990** [comment]
> What is still missing currently？

https://github.com/SpinalHDL/SpinalHDL/tree/dev/tester/src/test/scala/spinal/tester/scalatest is still full of things :)

**Comment by Readon** [comment]
> > What is still missing currently？
> 
> https://github.com/SpinalHDL/SpinalHDL/tree/dev/tester/src/test/scala/spinal/tester/scalatest is still full of things :)

ahh, that is only remained for the balance for running time of ci test.
